src/DotVVM.Framework/Controls/TextBox.cs

```diff
@@ -125,11 +125,21 @@ public static bool NeedsFormatting(IValueBinding binding) => binding != null &&
 
         protected internal override void OnPreRender(IDotvvmRequestContext context)
         {
-            isFormattingRequired = !string.IsNullOrEmpty(FormatString) ||
-                #pragma warning disable
+            if (!string.IsNullOrEmpty(FormatString) && (Type == TextBoxType.Number || Type == TextBoxType.Date))
+            {
+                throw new NotSupportedException($"Property StringFormat cannot be used with Type set to '{ Type }'." +
+                    $" In this case browsers localize '{ Type }' themselves.");
+            }
+
+            // Don't format <input type="Number" ... browsers expect
+            // value in specific format and localization breaks it
+            isFormattingRequired =  Type != TextBoxType.Number &&
```

Comment:
I think you should handle `TextBoxType.Date` here as well.

Reply:
In case of `TextBoxType.Date`  formatting is **required**. But I will try to make the solution more generic.